src/wp-includes/formatting.php

```diff
@@ -1278,32 +1306,271 @@ function wp_check_invalid_utf8( $text, $strip = false ) {
+ *     'the byte � is never allowed in a UTF-8 string.' === wp_scrub_utf8( $invalid, true );
+ *
+ *     // Maximal subparts are replaced individually.
+ *     '.�.' === wp_scrub_utf8( ".\xC0." );              // C0 is never valid.
+ *     '.�.' === wp_scrub_utf8( ".\xE2\x8C." );          // Missing A3 at end.
+ *     '.��.' === wp_scrub_utf8( ".\xE2\x8C\xE2\x8C." ); // Maximal subparts replaced separately.
+ *     '.��.' === wp_scrub_utf8( ".\xC1\xBF." );         // Overlong sequence.
+ *     '.���.' === wp_scrub_utf8( ".\xED\xA0\x80." );    // Surrogate half.
+ *
+ * @see https://www.unicode.org/versions/Unicode16.0.0/core-spec/chapter-5/#G40630
+ *
+ * @since 6.9.0
+ *
+ * @param string $text String which is assumed to be UTF-8 but may contain invalid sequences of bytes.
+ * @return string Input text with invalid sequences of bytes replaced with the Unicode replacement character.
+ */
+function wp_scrub_utf8( $text ) {
+	static $use_fallback = null;
+
+	if ( null === $use_fallback ) {
+		$use_fallback = (
+			! extension_loaded( 'mbstring' ) ||
+			// Maximal subpart substitution introduced by php/php-src@04e59c916f12b322ac55f22314e31bd0176d01cb.
+			version_compare( PHP_VERSION, '8.1.6', '<' )
+		);
 	}
-	// We can't demand utf8 in the PCRE installation, so just return the string in those cases.
-	if ( ! $utf8_pcre ) {
-		return $text;
+
+	if ( $use_fallback ) {
+		return _wp_scrub_utf8_fallback( $text );
 	}
 
-	// phpcs:ignore WordPress.PHP.NoSilencedErrors.Discouraged -- preg_match fails when it encounters invalid UTF8 in $text.
-	if ( 1 === @preg_match( '/^./us', $text ) ) {
-		return $text;
+	/*
+	 * While it looks like setting the substitute character could fail,
+	 * the internal PHP code will never fail when provided a valid
+	 * code point as a number. In this case, there’s no need to check
+	 * its return value to see if it succeeded.
+	 */
+	$prev_replacement_character = mb_substitute_character();
+	mb_substitute_character( 0xFFFD );
+	$scrubbed = mb_scrub( $text, 'UTF-8' );
+	mb_substitute_character( $prev_replacement_character );
+
+	/*
+	 * In PHP 8.0, `mb_scrub()` only returns a valid string. Once WordPress
+	 * depends on PHP 8.0.0 or above, this check will not be necessary.
+	 */
+	return false === $scrubbed
+		? _wp_scrub_utf8_fallback( $text )
+		: $scrubbed;
+}
+
+/**
+ * Fallback mechanism for replacing invalid spans of UTF-8 bytes.
+ *
+ * By implementing a raw method here the code will behave in the same way on
+ * all installed systems, regardless of what extensions are installed.
+ *
+ * This function largely duplicates the logic from {@see _wp_is_valid_utf8_fallback()}
+ * but where that function returns `false` when encountering invalid bytes, this
+ * function translates the invalid sequence of bytes starting at that offset.
+ *
+ * Example:
+ *
+ *     'Pi�a' === _wp_scrub_utf8_fallback( mb_convert_encoding( 'Piña', 'Windows-1252', 'UTF-8' ) );
```

Comment:
The use of `mb_convert_encoding` in the example adds a bit of complexity. I also think this doesn't show the intended behavior. It's converting a UTF-8 string `Piña` from 'Windows-1252' into 'UTF-8', where it becomes `Pi�a` (assuming `mb_substitute_character(substitute_character: 0xFFFD)`). Then the scrubbing doesn't do anything because it's valid UTF-8.

I like this example that doesn't require another function call:

https://github.com/WordPress/wordpress-develop/blob/66f3fa0903eff09af52ffc8d5ad29e0d12a7dd67/src/wp-includes/formatting.php#L939-L941


```suggestion
 *     'B�ch' === _wp_scrub_utf8_fallback( "B\xFCch" ); // "Büch" in ISO-8859-1 contains invalid UTF-8 byte 0xFC.
```

`Ñu` (Gnu in Spanish, 😉 to GPL) could also be fun, but `ñ` tends to be a longer byte sequence I think.

Reply:
it may seem confusing, but `iconv()` and `mb_convert_encoding()` have backwards from/to args. here, it converts _from_ UTF-8 (which is what’s provided in the literal comment text) into `Windows-1252`, which produces a valid string.

my intent was to make it clearer that we’re taking something encoded in `Windows-1252` and scrubbing it. apparently that failed, though I credit PHP in part because of the way that two functions which do approximately the same thing swap their argument order.

Reply:
Indeed, you're right. I confused the to/from encoding argument order. All the more reason to eliminate that from the example 😆

Reply:
The conversion has been removed. While I hoped it would demonstrate how we were prepping the bytes, the actual byte has been interpolated into the string and I left a comment indicating why there is an `\xF1` there